Initial implementation for SAP connector (#1)
wso2-ballerina/module-sap: component/src/main/wso2/sap/Module.md

## Package overview

> Ballerina SAP Client Endpoint is used to connect Ballerina with SAP. With this SAO Client Endpoint, Ballerina can act as SAP Consumers and SAP Producers.

Reviewer: ```suggestion
Ballerina SAP Client Endpoint is used to connect Ballerina with SAP. With this SAP Client Endpoint, Ballerina can act as SAP Consumers and SAP Producers.
```

> Ballerina SAP Client Endpoint is used to connect Ballerina with SAP. With this SAO Client Endpoint, Ballerina can act as SAP Consumers and SAP Producers.

Reviewer 2: ```suggestion
The Ballerina SAP Client Endpoint is used to connect Ballerina with SAP. This SAO Client Endpoint enables Ballerina to act as SAP Consumers and Producers.
```

It has full IDoc and experimental BAPI support. It uses the SAP JCO library as the underlying framework to communicate with SAP. 

## Steps to Configure

>  * Install the Java JDK 1.8

Reviewer 2: ```suggestion
 * Install Java JDK 1.8.
```

 * Download the sapidoc3.jar and sapjco3.jar middleware libraries from the SAP support portal and copy those libraries 
   to the <BRE_HOME>/bre/lib/ directory.

>  * Download the native SAP JCo library and copy it to the system path. You need to select the system path applicable 

Reviewer 2: ```suggestion
 * Download the native SAP JCo library and copy it to the system path. You need to select the system path that is applicable 
```

>    to your operating system as described below.

Reviewer 2: ```suggestion
   to your Operating System as described below.
```

    
    <table class="tg">
      <tr>
        <td class="tg-yw4l">Linux 32-bit</td>
        <td class="tg-yw4l">Copy the Linux native SAP jcolibrary libsapjco3.so to <JDK_HOME>/jre/lib/i386/server.</td>
      </tr>
      <tr>
        <td class="tg-yw4l">Linux 64-bit</td>
        <td class="tg-yw4l">Copy the Linux native SAP jcolibrary libsapjco3.so to <JDK_HOME>/jre/lib/amd64.</td>
      </tr>
      <tr>
        <td class="tg-yw4l">Windows</td>
        <td class="tg-yw4l">Copy the Windows native SAP jcolibrary sapjco3.dll to <WINDOWS_HOME>/system32.
        </td>
      </tr>
    </table>
    

>  * Extract wso2-sap-<version>.zip and  Run the install.sh script to install the module.

Reviewer 2: ```suggestion
 * Extract the wso2-sap-<version>.zip file and execute the install.sh script to install the module.
```

>    You can uninstall the module by running uninstall.sh.

Reviewer 2: ```suggestion
   You can uninstall the module by executing the uninstall.sh script.
```

Building From the Source
==================================

> If you want to build Ballerina sap endpoint from the source code:

Reviewer 2: ```suggestion
Follow the steps below to build the Ballerina SAP endpoint from the source code:
```

> 1. Get a clone or download the source from this repository:

Reviewer 2: ```suggestion
1. Get a clone or download [the source](https://github.com/wso2-ballerina/module-sap).
```

>     [https://github.com/wso2-ballerina/module-sap](https://github.com/wso2-ballerina/module-sap)

Reviewer 2: ```suggestion
 
```

> 2. Create lib folder into the module-sap 

Reviewer 2: ```suggestion
2. Create a lib folder in the module-sap directory.
```

3. Download the sapidoc3.jar and sapjco3.jar middleware libraries from the SAP support portal and copy those 

>    libraries to the module-sap/lib

Reviewer 2: ```suggestion
   libraries to the module-sap/lib directory.
```

> 4. Run the following Maven command from the ballerina directory:

Reviewer 2: ```suggestion
4. Navigate to the <BALLERINA_HOME> directory and execute the following Maven command:
```

    mvn clean install

> 5. Extract the distribution created at `/component/target/wso2-sap-<version>.zip`. Run the install.{sh/bat} script to install the module.

Reviewer 2: ```suggestion
5. Extract the `/component/target/wso2-sap-<version>.zip` distribution. 
6. Execute either of the install.{sh/bat} scripts to install the module.
```

> You can uninstall the module by running uninstall.{sh/bat}.    

Reviewer 2: ```suggestion
You can uninstall the module by executing either of the uninstall.{sh/bat} scripts.    
```

    
   
## Samples

### Sample SAP client endpoint 

```ballerina

sap:ProducerConfig producerConfigs = {
    destinationName:"<The SAP gateway name>",
    sapClient:"<SAP client, for example, 001>",
    userName:"<The user logon>",
    password:"<The logon password>",
    asHost:"<The R/3 application server>",
    sysnr:"<SAP system number, for example, 01>",
    language:"<The logon language>"
};

sap:Producer sapProducer = new(producerConfigs);
```

#### Sample SAP client operations

> Following example demonstrates a way to publish a bapi to sap.

Reviewer 2: ```suggestion
The following example demonstrates how to publish a BAPI to SAP.
```

```ballerina
string qname = "_-DSD_-ROUTEACCOUNT_CORDER002";
int idocVersion = The version of IDoc;
xml idoc = xml `<IDoc Data>`; 
NOTE: Here, If Any tag in the IDoc start with "_-", then you have to intialise that tag and use it in the IDoc. 
      eg: <{{qname}}>...</{{qname}}>  

public function main() {
    var result = sapProducer-> sendIdocMessage(idocVersion, idoc);
    if (result is error) {
        io:println("Error: ", result.reason());
    } else {
        io:println("Response: ", result);
    }
}
````

> Following example demonstrates a way to publish a bapi to sap.

Reviewer 2: ```suggestion
The following example demonstrates how to publish a BAPI to SAP.
```
